subsys/mpsl/fem/nrf21540_gpio_spi/mpsl_fem_nrf21540_gpio_spi.c

```diff
@@ -0,0 +1,441 @@
+     //spi cpnfig
+#if !defined(CONFIG_PINCTRL)
+
+			.mosi_pin = {
+#if DT_NODE_HAS_PROP(MPSL_FEM_SPI_BUS, mosi_pin)
+                .p_port        = MPSL_FEM_SPI_GPIO_PORT_REG(mosi_gpios),
+                .port_no       = MPSL_FEM_SPI_GPIO_PORT_NO(mosi_gpios),
+                .port_pin      = MPSL_FEM_SPI_GPIO_PIN_NO(mosi_gpios),
+#else
+                .port_pin      = MPSL_FEM_GPIO_INVALID_PIN
+#endif
+			},
+
+			.miso_pin = {
+#if DT_NODE_HAS_PROP(MPSL_FEM_SPI_BUS, miso_pin)
+                .p_port        = MPSL_FEM_SPI_GPIO_PORT_REG(miso_gpios),
+                .port_no       = MPSL_FEM_SPI_GPIO_PORT_NO(miso_gpios),
+                .port_pin      = MPSL_FEM_SPI_GPIO_PIN_NO(miso_gpios),
+#else
+                .port_pin      = MPSL_FEM_GPIO_INVALID_PIN
+#endif
+			},
+
+			.sck_pin = {
+#if DT_NODE_HAS_PROP(MPSL_FEM_SPI_BUS, sck_pin)
+                .p_port        = MPSL_FEM_SPI_GPIO_PORT_REG(sck_gpios),
+                .port_no       = MPSL_FEM_SPI_GPIO_PORT_NO(sck_gpios),
+                .port_pin      = MPSL_FEM_SPI_GPIO_PIN_NO(sck_gpios),
+#else
+                .port_pin      = MPSL_FEM_GPIO_INVALID_PIN
+#endif
+			},
+
+#endif // !defined(CONFIG_PINCTRL)
+
+			.cs_pin = {
+#if DT_NODE_HAS_PROP(MPSL_FEM_SPI_BUS, cs_gpios)
+                .p_port        = MPSL_FEM_SPI_GPIO_PORT_REG(cs_gpios),
+                .port_no       = MPSL_FEM_SPI_GPIO_PORT_NO(cs_gpios),
+                .port_pin      = MPSL_FEM_SPI_GPIO_PIN_NO(cs_gpios),
+#else
+                .port_pin      = MPSL_FEM_GPIO_INVALID_PIN
+#endif
+			}
+        }
+	};
+
+	IF_ENABLED(CONFIG_HAS_HW_NRF_PPI,
+		   (err = ppi_channel_alloc(cfg.ppi_channels, ARRAY_SIZE(cfg.ppi_channels));));
+	IF_ENABLED(CONFIG_HAS_HW_NRF_DPPIC,
+		   (err = ppi_channel_alloc(cfg.dppi_channels, ARRAY_SIZE(cfg.dppi_channels));));
+	if (err) {
+		return err;
+	}
+
+	IF_ENABLED(CONFIG_HAS_HW_NRF_DPPIC,
+		   (err = egu_instance_alloc(&cfg.egu_instance_no);));
+	if (err) {
+		return err;
+	}
+
+	IF_ENABLED(CONFIG_HAS_HW_NRF_DPPIC,
+		   (err = egu_channel_alloc(cfg.egu_channels,
+					    ARRAY_SIZE(cfg.egu_channels),
+					    cfg.egu_instance_no);))
+	if (err) {
+		return err;
+	}
+
+	BUILD_ASSERT(
+	  (CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB == CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB_POUTA) ||
+	  (CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB == CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB_POUTB));
+#if DT_NODE_HAS_PROP(DT_NODELABEL(nrf_radio_fem), mode_gpios)
+	if (!IS_ENABLED(CONFIG_MPSL_FEM_NRF21540_RUNTIME_PA_GAIN_CONTROL)) {
+		gpio_flags_t mode_pin_flags = GPIO_OUTPUT_ACTIVE;
+
+		if (CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB == CONFIG_MPSL_FEM_NRF21540_TX_GAIN_DB_POUTA) {
+			mode_pin_flags = GPIO_OUTPUT_INACTIVE;
+		}
+
+		mode_pin_flags |= DT_GPIO_FLAGS(DT_NODELABEL(nrf_radio_fem), mode_gpios);
+
+		err = gpio_pin_configure(
+			DEVICE_DT_GET(MPSL_FEM_GPIO_PORT(mode_gpios)),
+			MPSL_FEM_GPIO_PIN_NO(mode_gpios),
+			mode_pin_flags);
+
+		if (err) {
+			return err;
+		}
+	}
+#endif
+
+#if defined(CONFIG_PINCTRL)
```

Comment:
This function has become so long that it's mind-boggling. Please at least extract SPI configuration to a separate helper function so that lines 227-259 and 319-356 boil down to a single call. If you have more ideas how to shorten the function without too much fuss - go ahead.